### PR TITLE
rgw/aws/build: AWS submodule packaging

### PR DESCRIPTION
**Pull request by yuvalif**
* in this PR the `src/aws-sdk-cpp/` submodule points to the `install_only_runtime` branch in my fork of the `aws-sdk-cpp` repo (https://github.com/yuvalif/aws-sdk-cpp.git). this is done temporarily until https://github.com/ceph/aws-sdk-cpp/pull/2 is merged.
* this PR is only draft, and used to figure out the packing parts of the original PR: https://github.com/ceph/ceph/pull/36062

to test locally:
* call "make-dist" on the ceph root directory to generate srpm
* use "mock" to build packages based on that:
```
mock -r fedora-32-x86_64 --rootdir=<mock root dir> --rebuild <srpm file>
```

### CODE REVIEW

**Comment by yuvalif** [comment]
not working yet, but getting closer...

shaman builds on latest code are [here](https://shaman.ceph.com/builds/ceph/wip-yuval-aws-packaging2/95f8b4a8f6dfab5b002344797f4e24de4c045d64/)

- ubuntu builds are failing on cmake pthread issue - don't think it is related to the change here
- opensuse build is failing because of lack of internet access during the build. this is related to the changes, since the aws-sdk invoking "git clone" during build time to fetch its dependencies.
  -  can we change the opensuse build to allow for internet access?
  - another option would be to modify the aws-sdk to use submodules instead of custom dependencies - but this would be a major change there
- both "default" and "crimson" centos builds are failing to find: "libblk.so" and "libcrc32.so". not sure if this is related to my changes or not (though these libraries are unrelated to the aws libs)
- the "notcmalloc" centos build is passing!

File: src/test/rgw/CMakeLists.txt [hunk 1]
```diff
@@ -222,8 +222,4 @@ if(WITH_RADOSGW_AWS_ENDPOINT)
   add_executable(unittest_rgw_aws test_rgw_aws.cc)
   target_include_directories(unittest_rgw_aws PRIVATE "${CMAKE_SOURCE_DIR}/src/aws-sdk-cpp/aws-cpp-sdk-core/include")
   target_link_libraries(unittest_rgw_aws ${rgw_libs} ${UNITTEST_LIBS})
-  install(TARGETS
```
Comment (prateek2211) [inline src/test/rgw/CMakeLists.txt]: don't we need this line to install the unit test binary?

Comment (yuvalif) [reply src/test/rgw/CMakeLists.txt]: we don't install unit tests. they are built and placed into build/bin so we can run them in the test environment, but they are not needed on the machine where we install ceph

**Comment by badone** [comment]
> not working yet, but getting closer...
> 
> shaman builds on latest code are [here](https://shaman.ceph.com/builds/ceph/wip-yuval-aws-packaging2/95f8b4a8f6dfab5b002344797f4e24de4c045d64/)
> 
>     * ubuntu builds are failing on cmake pthread issue - don't think it is related to the change here

There is an earlier error that may be contributing (I sent you an email). Search for 'AwsCFlags'.

> 
>     * opensuse build is failing because of lack of internet access during the build. this is related to the changes, since the aws-sdk invoking "git clone" during build time to fetch its dependencies.

I don't know anything about opensuse nor how we would enable internet access during a build. Perhaps create a tracker?

>       
>       * can we change the opensuse build to allow for internet access?
>       * another option would be to modify the aws-sdk to use submodules instead of custom dependencies - but this would be a major change there
> 
>     * both "default" and "crimson" centos builds are failing to find: "libblk.so" and "libcrc32.so". not sure if this is related to my changes or not (though these libraries are unrelated to the aws libs)

This is during the creation of the container only, the build actually seems to succeed. I'm looking at why the container may be failing and will update with any findings here once I have any.

> 
>     * the "notcmalloc" centos build is passing!

**Comment by badone** [comment]
The last master build to succeed at the time I looked at this was https://shaman.ceph.com/builds/ceph/master/1897b8cf57936a432a7527650ff5e79c963b1f1f/  I rebased this PR on that commit and rebuilt at https://shaman.ceph.com/builds/ceph/wip-badone-testing/bd31db675ccd4fbb822d373940b863c085782bdc I then downloaded the rpms from each build and compared them.

```
$ rpm -qpR ceph-common-16.0.0-8710.g1897b8cf.el8.x86_64.rpm |egrep '(libblk.so|libcrc32.so)'
$ rpm -qpR ceph-common-16.0.0-8712.gbd31db67.el8.x86_64.rpm |egrep '(libblk.so|libcrc32)'
libblk.so()(64bit)
libcrc32.so()(64bit)
```
So the commits in this PR somehow introduce a dependency on libblk.so()(64bit) and libcrc32.so()(64bit) which can't be satisfied.

In a centos:8 container.

```
# dnf -y install https://2.chacra.ceph.com/r/ceph/wip-badone-testing/bd31db675ccd4fbb822d373940b863c085782bdc/centos/8/flavors/default/noarch/ceph-release-1-0.el8.noarch.rpm
# dnf install ceph-common
...
Error: 
 Problem: conflicting requests
  - nothing provides libcrc32.so()(64bit) needed by ceph-common-2:16.0.0-8712.gbd31db67.el8.x86_64
  - nothing provides libblk.so()(64bit) needed by ceph-common-2:16.0.0-8712.gbd31db67.el8.x86_64
# dnf provides \*/libcrc32.so\* \*/libblk.so\*
Last metadata expiration check: 0:04:27 ago on Thu Jan  7 03:15:38 2021.
Error: No Matches found
```
You either need to remove these dependencies or find some way to satisfy them on centos8.

**Comment by yuvalif** [comment]
> libcrc32.so

looks like `libcrc32.so` and `libblk.so` are libraries that we are building. 
for crs32:
see: https://github.com/ceph/ceph/blob/master/src/common/CMakeLists.txt#L195
and: https://github.com/ceph/ceph/blob/master/src/CMakeLists.txt#L450
for blk:
see: https://github.com/ceph/ceph/blob/master/src/blk/CMakeLists.txt#L28

I don't see how this PR introduces these dependencies

**Comment by yuvalif** [comment]
> I don't know anything about opensuse nor how we would enable internet access during a build. Perhaps create a tracker?

if we cross build it using `mock` then we should add the `--enable-network` flag for that to work.
do you know where the code for the shaman builds is?

**Comment by yuvalif** [comment]
> There is an earlier error that may be contributing (I sent you an email). Search for 'AwsCFlags'.

this looks like this [issue](https://github.com/aws/aws-sdk-cpp/issues/1384)
they said there,these failures are missing 3rd party dependencies.

however, in my [aws-sdk-cpp cmake file](https://github.com/ceph/aws-sdk-cpp/pull/2/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR74) the `BUILd_DEPS` flag is set to `ON`,
and according to [this](https://github.com/aws/aws-sdk-cpp#third-party-dependencies), it should be sufficient.

**Comment by badone** [comment]
> > libcrc32.so
> 
> looks like `libcrc32.so` and `libblk.so` are libraries that we are building.
> for crs32:
> see: https://github.com/ceph/ceph/blob/master/src/common/CMakeLists.txt#L195
> and: https://github.com/ceph/ceph/blob/master/src/CMakeLists.txt#L450
> for blk:
> see: https://github.com/ceph/ceph/blob/master/src/blk/CMakeLists.txt#L28
> 
> I don't see how this PR introduces these dependencies

Can you reproduce it without this PR?

**Comment by badone** [comment]
> > I don't know anything about opensuse nor how we would enable internet access during a build. Perhaps create a tracker?
> 
> if we cross build it using `mock` then we should add the `--enable-network` flag for that to work.
> do you know where the code for the shaman builds is?

https://github.com/ceph/ceph-build I believe. It currently does not use mock. Did you create the tracker?

**Comment by badone** [comment]
> > There is an earlier error that may be contributing (I sent you an email). Search for 'AwsCFlags'.
> 
> this looks like this [issue](https://github.com/aws/aws-sdk-cpp/issues/1384)
> they said there,these failures are missing 3rd party dependencies.
> 
> however, in my [aws-sdk-cpp cmake file](https://github.com/ceph/aws-sdk-cpp/pull/2/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR74) the `BUILd_DEPS` flag is set to `ON`,
> and according to [this](https://github.com/aws/aws-sdk-cpp#third-party-dependencies), it should be sufficient.

Did you try doing the build manually on Ubuntu?
> 
> also, the same code seems to work when building on centos.

That describes a lot of software doesn't it (builds on one distro, doesn't build on another)?

**Comment by badone** [comment]
> > libcrc32.so
> 
> looks like `libcrc32.so` and `libblk.so` are libraries that we are building.

Actually no, without the commits from this PR we build libcrc32.a and libblk.a and link statically (you can see the difference in 'ldd' output with, and without these commits). It's only in this PR we create `libcrc32.so` and `libblk.so` AFAICT.

> I don't see how this PR introduces these dependencies

It's pretty clear to me that it does.

**Comment by yuvalif** [comment]
> > > I don't know anything about opensuse nor how we would enable internet access during a build. Perhaps create a tracker?
> > 
> > 
> > if we cross build it using `mock` then we should add the `--enable-network` flag for that to work.
> > do you know where the code for the shaman builds is?
> 
> https://github.com/ceph/ceph-build I believe. It currently does not use mock. Did you create the tracker?

tracker created: https://tracker.ceph.com/issues/48817

**Comment by yuvalif** [comment]
> > > libcrc32.so
> > 
> > 
> > looks like `libcrc32.so` and `libblk.so` are libraries that we are building.
> 
> Actually no, without the commits from this PR we build libcrc32.a and libblk.a and link statically (you can see the difference in 'ldd' output with, and without these commits). It's only in this PR we create `libcrc32.so` and `libblk.so` AFAICT.
> 
> > I don't see how this PR introduces these dependencies
> 
> It's pretty clear to me that it does.

you are right. this is because the aws-sdk-cpp is setting a cmake var called `BUILD_SHARED_LIBS` causing our code to build shared libs instead of static libs.
trying to modify the aws-sdk-cpp cmake var to a different name: https://github.com/ceph/aws-sdk-cpp/pull/2/commits/4adc5dab5b16b199ef1046c75e38769d411ece4b

**Comment by yuvalif** [comment]
> > > There is an earlier error that may be contributing (I sent you an email). Search for 'AwsCFlags'.
> > 
> > 
> > this looks like this [issue](https://github.com/aws/aws-sdk-cpp/issues/1384)
> > they said there,these failures are missing 3rd party dependencies.
> > however, in my [aws-sdk-cpp cmake file](https://github.com/ceph/aws-sdk-cpp/pull/2/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR74) the `BUILd_DEPS` flag is set to `ON`,
> > and according to [this](https://github.com/aws/aws-sdk-cpp#third-party-dependencies), it should be sufficient.
> 
> Did you try doing the build manually on Ubuntu?
> 

yes. build without any issue on a ubuntu20 vm

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
this draft PR is now closed in favor of : https://github.com/ceph/ceph/pull/41439

**Pull request closed without merging**